src/ch/ch_monitor.c

```diff
@@ -560,6 +564,334 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+    event = virJSONValueObjectGetString(eventJSON, "event");
+
+    ev = virCHMonitorEventFromString(source, event);
+    VIR_DEBUG("Source: %s Event: %s, ev: %d", source, event, ev);
+    switch (ev) {
+        case virCHMonitorVmEventBooted:
+        case virCHMonitorVmEventResumed:
+        case virCHMonitorVmEventRestored:
+        case virCHMonitorVirtioDeviceEventActivated:
+        case virCHMonitorVirtioDeviceEventReset:
+            virObjectLock(vm);
+            if (virDomainObjIsActive(vm) &&
+                   virCHDomainObjBeginJob(vm, CH_JOB_MODIFY) == 0) {
+                virCHProcessSetupThreads(vm);
+                virCHDomainObjEndJob(vm);
+            }
+            virObjectUnlock(vm);
+            break;
+        case virCHMonitorVmmEventShutdown: // shutdown inside vmm
+        case virCHMonitorVmEventShutdown:
+            {
+                virCHDriverPtr driver = CH_DOMAIN_PRIVATE(vm)->driver;
+                g_autoptr(virCHDriverConfig) cfg = virCHDriverGetConfig(driver);
+                virDomainState state;
+
+                virObjectLock(vm);
+                state = virDomainObjGetState(vm, NULL);
+                if ((ev == virCHMonitorVmmEventShutdown ||
+                     state == VIR_DOMAIN_SHUTDOWN) &&
+                      (virCHDomainObjBeginJob(vm, CH_JOB_MODIFY) == 0)) {
+
+                    if (virDomainObjSave(vm, driver->xmlopt, cfg->stateDir))
+                        VIR_WARN("Failed to persist the domain after shutdown!");
+
+                    virCHProcessStop(driver, vm, VIR_DOMAIN_SHUTOFF_SHUTDOWN);
+
+                    virCHDomainObjEndJob(vm);
+                }
+                virObjectUnlock(vm);
+                break;
+            }
+        case virCHMonitorVmEventBooting:
+        case virCHMonitorVmEventPausing:
+        case virCHMonitorVmEventPaused:
+        case virCHMonitorVmEventResuming:
+        case virCHMonitorVmEventSnapshotting:
+        case virCHMonitorVmEventSnapshotted:
+        case virCHMonitorVmEventRestoring:
+        case virCHMonitorVmEventResizing:
+        case virCHMonitorVmEventResized:
+        case virCHMonitorVmEventDeleted:
+        case virCHMonitorVmmEventStarting:
+        case virCHMonitorCpuCreateVcpu:
+            break;
+        case virCHMonitorEventMax:
+        default:
+            VIR_WARN("unkown event from monitor!");
+            break;
+    }
+
+    return 0;
+}
+
+/*
+ * Helper function to find a block of valid JSON
+ * from a stream of multiple JSON blocks.
+ */
+static inline char *end_of_json(char *str, size_t len)
+{
+    bool started = false;
+    int blocks = 0;
+    int i = 0;
+
+    while ((i < len) && (!started || blocks > 0)) {
+        if (str[i] == '{') {
+            if (!started)
+                started = true;
+            blocks++;
+        } else if (str[i] == '}') {
+            blocks--;
+        }
+        i++;
+    }
+
+    return (i == len && blocks) ? NULL : str + i;
```

Comment:
same here, can you use the `virJSONValue*` API instead ?

Reply:
Same  as above. We need to split into individual JSON documents and then pass it on to virJSONValueFromString one by one.